File: android/src/main/java/expo/modules/applemusic/AppleMusicErrors.kt
package expo.modules.applemusic

import expo.modules.kotlin.exception.CodedException

internal object AppleMusicErrors {
  fun missingMusicUserToken(): CodedException =
    CodedException(
      AppleMusicErrorCodes.MISSING_MUSIC_USER_TOKEN,
      "Apple Music music user token is required. Call Auth.authorize(developerToken) first, then pass the token from AuthorizeResult on user-scoped APIs (or queue playback after authorize on Android).",
      null,
    )

  fun missingTokens(): CodedException =
    CodedException(
      AppleMusicErrorCodes.PERMISSION_DENIED,
      "Apple Music authorization required. Call Auth.authorize() first.",
      null,
    )

  fun permissionDenied(): CodedException =
    CodedException(
      AppleMusicErrorCodes.PERMISSION_DENIED,
      "Apple Music authorization required or subscription needed (403)",
      null,
    )

  fun apiError(message: String, code: String = AppleMusicErrorCodes.ERROR): CodedException =
    CodedException(code, message, null)

  fun playlistNotFound(): CodedException =
    CodedException(AppleMusicErrorCodes.ERROR, "Playlist not found in library", null)

  fun itemNotFound(item: String, inLibrary: Boolean): CodedException {
    val source = if (inLibrary) "library" else "catalog"
    return CodedException(AppleMusicErrorCodes.ERROR, "$item not found in $source", null)
  }

  fun unknownMediaType(type: String): CodedException =
    CodedException(AppleMusicErrorCodes.ERROR, "Unknown media type: $type", null)

  fun unsupportedLibraryType(type: String): CodedException =
    CodedException(AppleMusicErrorCodes.ERROR, "Unsupported library media type: $type", null)

  fun noSongsInPlaylist(): CodedException =
    CodedException(AppleMusicErrorCodes.ERROR, "No songs in playlist", null)
}

File: src/web/apple-music-errors.ts
import { CodedError } from 'expo-modules-core';
import { AppleMusicErrorCode } from '../constants/apple-music-error-codes';

export function missingDeveloperToken(): CodedError {
  return new CodedError(
    AppleMusicErrorCode.missingDeveloperToken,
    'Apple Music developer token is required on web. Pass a signed JWT to Auth.authorize(token).',
  );
}

export function missingTokens(): CodedError {
  return new CodedError(
    AppleMusicErrorCode.permissionDenied,
    'Apple Music authorization required. Call Auth.authorize() first.',
  );
}

export function permissionDenied(): CodedError {
  return new CodedError(
    AppleMusicErrorCode.permissionDenied,
    'Apple Music authorization required or subscription needed (403)',
  );
}

export function apiError(message: string, code = AppleMusicErrorCode.error): CodedError {
  return new CodedError(code, message);
}

export function playlistNotFound(): CodedError {
  return new CodedError(AppleMusicErrorCode.error, 'Playlist not found in library');
}

export function itemNotFound(item: string, inLibrary: boolean): CodedError {
  const source = inLibrary ? 'library' : 'catalog';
  return new CodedError(AppleMusicErrorCode.error, `${item} not found in ${source}`);
}

export function unknownMediaType(type: string): CodedError {
  return new CodedError(AppleMusicErrorCode.error, `Unknown media type: ${type}`);
}

export function unsupportedLibraryType(type: string): CodedError {
  return new CodedError(AppleMusicErrorCode.error, `Unsupported library media type: ${type}`);
}

export function noSongsInPlaylist(): CodedError {
  return new CodedError(AppleMusicErrorCode.error, 'No songs in playlist');
}
